### PR TITLE
Expression language

### PR DESCRIPTION
**Pull request by mandarjog**
Introducing expression language.

Front end uses golang parser.
Internal AST only has 3 nodes: `variables, constants, functions`
All of them are directly or indirectly typed. All operators are functions. 
`substring(a, 5) == "abc"  --> EQ(substring($a, 5), "abc")`

Supports following types of expressions.  Supports predicates and other expression in a uniform way.

Predicates   returns bool
```
service.name == "*.ns1.cluster" && service.user == "admin"
request.user == "user1"
(x == 20 && y == 10) || x == 30

request.header["X-FORWARDED-HOST"] == "aaa"`
```
Mapping Expressions
```
request.user2| request.user3 | "user1"     --> returns string
`request.size| 200`                                    --> returns int
```

Functions are implemented using `Func` interface
```
type Func interface {
        // Name uniquely identified the function.
        Name() string
        // ReturnType specifies the return type of this function.
        ReturnType() config.ValueType
        // ArgTypes specifies the argument types in order expected by the function.
        ArgTypes() []config.ValueType
        // NullArgs specifies if the function accepts null args
        NullArgs() bool
        // Call performs the function call. It is guaranteed
        // that call will be made with correct types and arity.
        // may panic.
        Call([]interface{}) interface{}
}
```

All expression can be type checked during validation.

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/255)

### CODE REVIEW

File: pkg/expr/expr.go [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+	default: // string
+		if typedVal, err = strconv.Unquote(v); err != nil {
+			return nil, err
+		}
+	}
+	return &Constant{Value: v, Kind: vType, TypedValue: typedVal}, nil
+}
+
+// Constant models a typed constant.
+type Constant struct {
```
Comment (ZackButcher) [inline pkg/expr/expr.go]: The field's names feel off here; I think it'd make more sense as:

    struct Constant struct {
        Str String
        Value interface{}
        Type config.ValueType
    }

File: pkg/expr/expr.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+		w.WriteString(arg.String())
+	}
+	w.WriteString(")")
+	return w.String()
+}
+
+// Eval evaluate function.
+func (f *Function) Eval(attrs attribute.Bag, fMap map[string]Func) (interface{}, error) {
+	fn := fMap[f.Name]
+	if fn == nil {
```
Comment (ZackButcher) [inline pkg/expr/expr.go]: fn, found := fMap[f.Name]
    if !found {
        ...
    }

is more idiomatic

File: pkg/expr/expr.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+		return nil, fmt.Errorf("unknown function: %s", f.Name)
+	}
+	// can throw NPE if config is not consistent.
```
Comment (ZackButcher) [inline pkg/expr/expr.go]: -> will return an error for nil args if config is not consistent

File: pkg/expr/func.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	return res
+}
+
+func (f *eqFunc) call(args []interface{}) bool {
+	var s0 string
+	var s1 string
+	var ok bool
+
+	if s0, ok = args[0].(string); !ok {
+		return args[0] == args[1]
```
Comment (ZackButcher) [inline pkg/expr/func.go]: This doesn't work for cases where args are things like `[]byte`; I think you might need to use [`reflect.DeepEqual`](https://golang.org/pkg/reflect/#DeepEqual) here.

File: pkg/expr/func.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+func (f *baseFunc) NullArgs() bool               { return f.nullArgs }
+
+type eqFunc struct {
+	*baseFunc
+	invert bool
+}
+
+// newEQFunc returns dynamically typed equality fn.
+// treats strings specially with prefix and suffix '*' matches
+func newEQFunc() Func {
```
Comment (ZackButcher) [inline pkg/expr/func.go]: nit: I don't think the `Func` suffix is necessary on all of these functions. `newEQ`, `newLOR`, etc. feel fine to me.

File: pkg/expr/func.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+	*baseFunc
+}
+
+// newIndexFunc returns a map accessor fn.
+func newIndexFunc() Func {
+	return &indexFunc{
+		baseFunc: &baseFunc{
+			name:     "INDEX",
+			retType:  config.VALUE_TYPE_UNSPECIFIED,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.STRING},
```
Comment (ZackButcher) [inline pkg/expr/func.go]: We should just add a `MAP_...` type to the `ValueType` proto if we're doing this.

File: pkg/expr/expr.go [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+	return nil
+}
+
+// Parse parses a given expression to ast.Expression.
+func Parse(src string) (ex *Expression, err error) {
+	a, err := parser.ParseExpr(src)
+	if err != nil {
+		return nil, fmt.Errorf("parse error: %s %s", src, err)
+	}
+	glog.V(2).Infof("\n\n%s : %#v\n", src, a)
```
Comment (ZackButcher) [inline pkg/expr/expr.go]: `"\n\n%s : %#v\n"` => `parsed '%s' into '%#v'`

glog handles newlines for us and the line as written doesn't provide any context

File: pkg/expr/func.go (hunk 4 above)
Comment (mandarjog) [inline pkg/expr/func.go]: good catch.

File: pkg/expr/func.go (hunk 6 above)
Comment (mandarjog) [inline pkg/expr/func.go]: I think it has already been done.

Comment (ZackButcher) [inline pkg/expr/func.go]: Shouldn't it be used here then? The first arg must be a map, not an unspecified value.

**Comment by mandarjog** [comment]
It is a subset of golang expressions that is common to other c like languages. (javascript etc.)
The only thing that it forbids is a python like  `in` operator.
If we really need that we will implement a new parser.

The backend will still implement it as `in (needle, haystack ...)`

File: pkg/expr/func.go (hunk 6 above)
Comment (mandarjog) [inline pkg/expr/func.go]: ok, done.

File: pkg/expr/func.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+	Name() string
+
+	// ReturnType specifies the return type of this function.
+	ReturnType() config.ValueType
+
+	// ArgTypes specifies the argument types in order expected by the function.
+	ArgTypes() []config.ValueType
+
+	// NullArgs specifies if the function accepts null args
+	NullArgs() bool
```
Comment (douglas-reid) [inline pkg/expr/func.go]: nit: would `AcceptsNulls` be a more descriptive name?

File: pkg/expr/expr.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+	return false, fmt.Errorf("typeError: got %s, expected bool", reflect.TypeOf(uret).String())
+}
+
+// Validate validates expression for syntactic correctness.
+// TODO check if all functions and attributes in the expression are defined.
+// at present this violates the contract with Func.Call that ensures
+// arity and arg types. It is upto the policy author to write correct policies.
+func (e *cexl) Validate(s string) (err error) {
+	var ex *Expression
+	if ex, err = Parse(s); err != nil {
```
Comment (douglas-reid) [inline pkg/expr/expr.go]: nit: this might read easier without the named return param...

```
ex, err := Parse(s)
if err != nil {
    return err
}
...
```

File: pkg/expr/func.go [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+		return strings.HasSuffix(s0, s1[1:])
+	}
+	return s0 == s1
+}
+
+type lAndFunc struct {
+	*baseFunc
+}
+
+// newLANDFunc returns a binary logical AND fn.
```
Comment (douglas-reid) [inline pkg/expr/func.go]: nit: looks like this is now called just `newLAND`

File: pkg/expr/func.go [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+func (f *lAndFunc) Call(args []interface{}) interface{} {
+	return args[0].(bool) && args[1].(bool)
+}
+
+type lOrFunc struct {
+	*baseFunc
+}
+
+// newLORFunc returns logical OR fn.
+func newLOR() Func {
```
Comment (douglas-reid) [inline pkg/expr/func.go]: nit: just `newLOR` in comment.

File: pkg/expr/func.go [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+		&baseFunc{
+			name:     "LOR",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.BOOL, config.BOOL},
+			nullArgs: false,
+		},
+	}
+}
+
+// Call should return thru if at least one element is true
```
Comment (douglas-reid) [inline pkg/expr/func.go]: should return `true` ?

File: pkg/expr/func.go [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+// applies to non bools.
+type orFunc struct {
+	*baseFunc
+}
+
+// newORFunc returns an OR fn that selects first non empty argument.
+// types can be anything, but all args must be of the same type.
+func newOR() Func {
```
Comment (douglas-reid) [inline pkg/expr/func.go]: comment -> just `newOR`

File: pkg/expr/eval_test.go [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+			},
+			true, "unresolved attribute",
+		},
+	}
+
+	for idx, tst := range tests {
+		attrs := &bag{attrs: tst.tmap}
+		exp, err := Parse(tst.src)
+		if err != nil {
+			t.Errorf("[%d] unexpected error: %s", idx, err)
```
Comment (mandarjog) [inline pkg/expr/eval_test.go]: good point, done.
It seems to have caught up with Gomega tests.

File: pkg/expr/func.go (hunk 8 above)
Comment (mandarjog) [inline pkg/expr/func.go]: done

File: pkg/expr/func.go (hunk 10 above)
Comment (mandarjog) [inline pkg/expr/func.go]: one.

File: pkg/expr/func.go (hunk 11 above)
Comment (mandarjog) [inline pkg/expr/func.go]: done

File: pkg/expr/func.go (hunk 13 above)
Comment (mandarjog) [inline pkg/expr/func.go]: done.

**Comment by mandarjog** [comment]
PTAL.
Expression language is in and type validation is also implemented.

File: pkg/expr/expr_test.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+	config "istio.io/api/mixer/v1/config/descriptor"
+	dpb "istio.io/api/mixer/v1/config/descriptor"
+)
+
+func TestGoodParse(tst *testing.T) {
+	tests := []struct {
+		src     string
+		postfix string
+	}{
+		{"a.b.c == 2", "EQ($a.b.c, 2)"},
```
Comment (wora) [inline pkg/expr/expr_test.go]: I prefer @a.b.c because @ is close to attribute than $.

File: pkg/expr/expr_test.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+		{`!(a.b || b || "c a b" || ( a && b ))`, `NOT(LOR(LOR(LOR($a.b, $b), "c a b"), LAND($a, $b)))`},
```
Comment (wora) [inline pkg/expr/expr_test.go]: I wonder if we can use operators directly, such as `a + b` => `+(a, b)`. In this way, we can allow more operators without having to define a large grammar.

File: bin/fmt.sh [hunk 17]
```diff
@@ -3,13 +3,24 @@
...
 
 set -e
-
 SCRIPTPATH=$( cd "$(dirname "$0")" ; pwd -P )
 ROOTDIR=$SCRIPTPATH/..
 cd $ROOTDIR
 
-gofmt -s -w adapter cmd pkg
-goimports -w adapter cmd pkg
+GO_FILES=$(find adapter cmd pkg -type f -name '*.go')
```
Comment (douglas-reid) [inline bin/fmt.sh]: nit: moving forward, let's separate out unrelated changes into distinct PRs (keep them small, reduce load, etc.)

File: cmd/server/server.go [hunk 18]
```diff
@@ -134,7 +134,7 @@ func serverOpts(sa *serverArgs, handlers api.Handler) (*api.GRPCServerOptions, e
 
 func runServer(sa *serverArgs) error {
 	// get aspect registry with proper aspect --> api mappings
-	eval := expr.NewIdentityEvaluator()
+	eval := expr.NewCexlEvaluator()
```
Comment (douglas-reid) [inline cmd/server/server.go]: is Cexl an initialism?  If so, should this be `NewCEXLEvaluator` ?

File: pkg/attribute/definitionFinder.go [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+// limitations under the License.
+
+package attribute
+
+import (
+	dpb "istio.io/api/mixer/v1/config/descriptor"
+)
+
+// DefinitionFinder finds attribute definitions.
+type DefinitionFinder interface {
```
Comment (douglas-reid) [inline pkg/attribute/definitionFinder.go]: Since the method is `FindAttribute`, should this be `AttributeFinder` ? 

Would `DescriptorFinder` and `FindDescriptor` perhaps be more appropriate?

File: pkg/config/runtime.go [hunk 20]
```diff
@@ -87,7 +88,7 @@ func (r *Runtime) resolveRules(bag attribute.Bag, aspectSet AspectSet, rules []*
 	var lerr error
 
 	for _, rule := range rules {
-		if glog.V(2) {
+		if glog.V(3) {
```
Comment (douglas-reid) [inline pkg/config/runtime.go]: curious: why do we use this form here, and the inline form below?  is rule that much of a pain to stringify in comparison?

File: pkg/expr/eval_test.go (hunk 14 above)
Comment (douglas-reid) [inline pkg/expr/eval_test.go]: just as an FYI: I don't see a change here to the tests. not a blocker, but I'm wondering if there is a missing commit ?

File: pkg/expr/eval_test.go [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+			continue
+		}
+		if res != tst.result {
+			t.Errorf("[%d] got %s\nwant %s", idx, res, tst.result)
+		}
+	}
+
+}
+
+func TestCexlEval(t *testing.T) {
```
Comment (douglas-reid) [inline pkg/expr/eval_test.go]: testCEXLEval ?

File: pkg/expr/eval_test.go [hunk 22]
```diff
@@ -0,0 +1,375 @@
...
+			continue
+		}
+		// check if error is of the correct type
+		if err != nil && !strings.Contains(err.Error(), tt.err) {
+			t.Errorf("[%d] got %s, want %s", idx, err, tt.err)
+		}
+	}
+}
+
+// fake bag
```
Comment (douglas-reid) [inline pkg/expr/eval_test.go]: aside: i wonder if we should create an pkg/attribute/bagtest.go or something. we have several test impls, including pkg/aspect/test/bag.go throughout the codebase that we should try to harmonize.

File: pkg/expr/func.go [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+package expr
+
+import (
+	"reflect"
+	"strings"
+
+	config "istio.io/api/mixer/v1/config/descriptor"
+)
+
+// Func every expression function provider should have this interface.
```
Comment (douglas-reid) [inline pkg/expr/func.go]: nit: make this a complete sentence. something like: `Func defines the interface that every expression function provider must implement`.

File: testdata/globalconfig.yml [hunk 24]
```diff
@@ -6,12 +6,8 @@ adapters:
     impl: memQuota
     params:
   - name: default
-    kind: access-logs
```
Comment (douglas-reid) [inline testdata/globalconfig.yml]: are these changes related to the previous refactors?

File: pkg/expr/eval_test.go (hunk 14 above)
Comment (mandarjog) [inline pkg/expr/eval_test.go]: I just used it in new tests in expr_test.go

File: pkg/attribute/definitionFinder.go (hunk 19 above)
Comment (mandarjog) [inline pkg/attribute/definitionFinder.go]: fixed.

File: cmd/server/server.go (hunk 18 above)
Comment (mandarjog) [inline cmd/server/server.go]: done.

File: bin/fmt.sh (hunk 17 above)
Comment (mandarjog) [inline bin/fmt.sh]: 👍

File: pkg/config/runtime.go (hunk 20 above)
Comment (mandarjog) [inline pkg/config/runtime.go]: I think at one point there was another statement inside glog.V(2). Changes to inline form now.

File: pkg/expr/eval_test.go (hunk 21 above)
Comment (mandarjog) [inline pkg/expr/eval_test.go]: done.

File: pkg/expr/eval_test.go (hunk 22 above)
Comment (mandarjog) [inline pkg/expr/eval_test.go]: https://github.com/istio/mixer/issues/263

File: pkg/expr/expr_test.go (hunk 16 above)
Comment (mandarjog) [inline pkg/expr/expr_test.go]: function names have to be valid identifiers. So that excludes any operators as function names. 
I would like to minimize number of operators beyond the customary ones that we must support.

File: pkg/expr/func.go (hunk 23 above)
Comment (mandarjog) [inline pkg/expr/func.go]: done.

File: testdata/globalconfig.yml (hunk 24 above)
Comment (mandarjog) [inline testdata/globalconfig.yml]: Yes. `kind` is specified by the adapter implementation now. It is no longer needed  (allowed) in the global config.

**Comment by mandarjog** [comment]
PTAL.

File: pkg/expr/func.go [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+	// Name uniquely identified the function.
+	Name() string
+
+	// ReturnType specifies the return type of this function.
+	ReturnType() config.ValueType
+
+	// ArgTypes specifies the argument types in order expected by the function.
+	ArgTypes() []config.ValueType
+
+	// NullArgs specifies if the function accepts null args
```
Comment (douglas-reid) [inline pkg/expr/func.go]: nit: comment no longer matches method name.
